### PR TITLE
planner, executor: fix index merge partial table scan schema

### PR DESCRIPTION
**Pull request by tangenta**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/23919

Problem Summary: 

```sql
create table t (a int, b int, index(a), index(b)) partition by hash (a) partitions 2;
insert into t values (1, 5);
select /*+ use_index_merge( t ) */ * from t where a in (3) or b in (5) order by a;
```

```console
ERROR 1105 (HY000): runtime error: index out of range [2] with length 2
```

The explain result:
```
+--------------------------------------+---------+-----------+-----------------------------------+---------------------------------------------+
| id                                   | estRows | task      | access object                     | operator info                               |
+--------------------------------------+---------+-----------+-----------------------------------+---------------------------------------------+
| Sort_9                               | 39.98   | root      |                                   | test.t.a                                    |
| └─PartitionUnion_12                  | 39.98   | root      |                                   |                                             |
|   ├─IndexMerge_16                    | 19.99   | root      |                                   |                                             |
|   │ ├─IndexRangeScan_13(Build)       | 10.00   | cop[tikv] | table:t, partition:p0, index:a(a) | range:[3,3], keep order:false, stats:pseudo |
|   │ ├─IndexRangeScan_14(Build)       | 10.00   | cop[tikv] | table:t, partition:p0, index:b(b) | range:[5,5], keep order:false, stats:pseudo |
|   │ └─TableRowIDScan_15(Probe)       | 19.99   | cop[tikv] | table:t, partition:p0             | keep order:false, stats:pseudo              |
|   └─IndexMerge_20                    | 19.99   | root      |                                   |                                             |
|     ├─IndexRangeScan_17(Build)       | 10.00   | cop[tikv] | table:t, partition:p1, index:a(a) | range:[3,3], keep order:false, stats:pseudo |
|     ├─IndexRangeScan_18(Build)       | 10.00   | cop[tikv] | table:t, partition:p1, index:b(b) | range:[5,5], keep order:false, stats:pseudo |
|     └─TableRowIDScan_19(Probe)       | 19.99   | cop[tikv] | table:t, partition:p1             | keep order:false, stats:pseudo              |
+--------------------------------------+---------+-----------+-----------------------------------+---------------------------------------------+
```

`SortExec.Next()` only requires 2 columns(`a` and `b`), but the row fetched from the underlying `UnionExec` has 3(`a`, `b` and `_tidb_rowid`). This cause `Chunk.AppendPartialRow` panic because the slice out-of-range error.

The request chunk for `UnionExec.Next` should have only 2 columns, but `req.SwapColumns(result.chk)` change it into a 3-columns chunk. This chunk comes from `IndexMergeReaderExecutor`, which has a 3-columns schema.

`IndexMergeReaderExecutor` inherits the schema from `PhysicalIndexMergeReader`, which inherits from the `PhysicalTableScan` built by `buildIndexMergeTableScan`:

https://github.com/pingcap/tidb/blob/008b91b12b0780539911da9f533d42114a213d87/planner/core/find_best_task.go#L903-L914

It appends the `_tidb_rowid` to the schema of `DataSource` to help `HandleCols` resolve its indices.

The `HandleCols` is used to fetch handles in both `TableRangeScan` and `IndexRangeScan`. 
- For the `TableRangeScan`, the current schema is from `DataSource` plus additional common handle columns. This is unresonable because only the handle columns are needed.
- For the `IndexRangeScan`, the current schema is always indexed columns plus full common handle columns. `ResolveIndices` is useless in `IndexRangeScan` because the `expression.Column.Index` is never used.

### What is changed and how it works?

How it Works: 

This PR makes the schema of `IndexMerge` consistent with the upper operator, by avoid changing `DataSource` schema directly. It also fixes the issue in https://github.com/pingcap/tidb/pull/23933#pullrequestreview-633357194.

What is changed:

- Change the schema of `TableRangeScan`(as a child of `IndexMerge` operator) to the one that only contains handle columns.
- Resolve the `HandleCols` in `IndexMergeReaderExecutor` against the schema that only contains the handle columns.
- Remove unnecessary projection.
- Add test for #23919 and `clustered index/_tidb_rowid/PkIsHandle` for index merge.

### Related changes

 NA

### Check List 

Tests 

- Unit test
- Integration test

Side effects

NA

### Release note 

- No release note

### CODE REVIEW

**Review by lzmhhh123 — APPROVED** [review]

File: planner/core/find_best_task.go [hunk 1]
```diff
@@ -888,20 +882,11 @@ func (ds *DataSource) convertToPartialIndexScan(prop *property.PhysicalProperty,
...
-			if ts.schema.ColumnIndex(col) == -1 {
-				ts.Schema().Append(col)
-				ts.Columns = append(ts.Columns, col.ToInfo())
-				needExtraProj = true
-			}
-		}
-	}
 	rowSize := ds.TblColHists.GetAvgRowSize(ds.ctx, ds.TblCols, false, false)
 	sessVars := ds.ctx.GetSessionVars()
+	overwritePartialTableScanSchema(ds, ts)
```
Comment (eurekaka) [inline planner/core/find_best_task.go:889]: It's better to update the `rowSize` as well since we only return the handle column from the coprocessor to tidb now, for a more accurate cost estimation. You can do it in another PR if you prefer of course.

**Review by eurekaka — APPROVED** [review]
/lgtm

**Review by eurekaka — COMMENTED** [review]
/lgtm
